src/MetadataUtility/Audio/Wave.cs

```diff
@@ -2,12 +2,156 @@
 // All code in this file and all associated files are the copyright and property of the QUT Ecoacoustics Research Group.
 // </copyright>
 
-namespace MetadataUtility.Audio.Vendors
+namespace MetadataUtility.Audio
 {
+    using LanguageExt;
+    using LanguageExt.Common;
+    using MetadataUtility.Utilities;
+
     public static class Wave
     {
+        public const string Mime = "audio/wave";
+
+        public const int WaveFileOffset = 7;
+        public const int WaveSampleRateOffset = 24;
+        public const int WaveChannelOffset = 22;
+        public const int WaveBitsPerSampleOffset = 34;
+        public const int WaveFileLengthOffset = 4;
+
         public static readonly byte[] WaveMagicNumber = new byte[] { (byte)'W', (byte)'A', (byte)'V', (byte)'E' };
 
+        public static readonly Error FileTooShort = Error.New("Error reading file: file is not long enough to have a duration header");
+
+        public static readonly Error FileTooShortWave = Error.New("Error reading file: file is not long enough to have a WAVE header");
+
         public static readonly byte[] DataBlockId = new byte[] { (byte)'d', (byte)'a', (byte)'t', (byte)'a' };
+
+        public static Fin<ulong> ReadTotalSamples(FileStream stream)
+        {
+            stream.Seek(WaveFileOffset, SeekOrigin.Begin);
+
+            Span<byte> buffer = stackalloc byte[5];
+            int bytesRead = stream.Read(buffer);
+
+            if (bytesRead != buffer.Length)
+            {
+                return FileTooShort;
+            }
+
+            // flac files have an unsigned 36-bit integer for total sample duration!
+            return BinaryHelpers.Read36BitUnsignedBigEndianIgnoringFirstOctet(buffer);
+        }
+
+        public static Fin<ulong> ReadWaveSampleRate(FileStream stream)
+        {
+            stream.Seek(WaveSampleRateOffset, SeekOrigin.Begin);
+
+            Span<byte> buffer = stackalloc byte[4];
+            int bytesRead = stream.Read(buffer);
+
+            if (bytesRead != buffer.Length)
+            {
+                return FileTooShort;
+            }
+
+            if (buffer.Length < 4)
+            {
+                throw new ArgumentException("bytes span must at least be 4 long", nameof(buffer));
```

Comment:
so return one of our `Error` style errors here (like `TooShort`) rather than throwing

The idea is these method should never throw, and that failures should always have a useful error message.

If an exception were the right choice here, we'd probably choose `InvalidOperationException` over an `ArgumentException` since the problem is with how we're operating and not with an argument to the current method.